rust/emitter/src/reference_op_emitter.rs

```diff
@@ -0,0 +1,501 @@
+                //      [stack] VAL
+            }
+        }
+
+        Ok(())
+    }
+}
+
+// Struct for emitting bytecode for new operation.
+//
+// Usage:
+//
+//   `new expr(a);`
+//      NewEmitter::emit(self, |emitter| {
+//          emitter.emit_expression(expr)
+//      }, |emitter| {
+//          // Return the number of arguments.
+//          emitter.emit_expression(a)?;
+//          Ok(1)
+//      })?;
+//
+pub struct NewEmitter {}
+impl NewEmitter {
+    pub fn emit<F1, F2>(
+        emitter: &mut AstEmitter,
+        callee_emitter: F1,
+        arguments_emitter: F2,
+    ) -> Result<(), EmitError>
+    where
+        F1: Fn(&mut AstEmitter) -> Result<(), EmitError>,
+        F2: Fn(&mut AstEmitter) -> Result<usize, EmitError>,
+    {
+        //              [stack]
+
+        callee_emitter(emitter)?;
+        //              [stack] CALLEE
+
+        emitter.emit.is_constructing();
+        //              [stack] CALLEE JS_IS_CONSTRUCTING
+
+        // FIXME: Support spread.
+        let len = arguments_emitter(emitter)?;
+        //              [stack] CALLEE JS_IS_CONSTRUCTING ARGS...
+
+        emitter.emit.dup_at(len as u32 + 1);
+        //              [stack] CALLEE JS_IS_CONSTRUCTING ARGS... CALLEE
+
+        emitter.emit.new_(len as u16);
+        //              [stack] VAL
+
+        Ok(())
+    }
+}
+
+// See AssignmentEmitter.
+pub struct SetOpReferenceEmitter {}
+impl SetOpReferenceEmitter {
+    pub fn emit_name(emitter: &mut AstEmitter, name: &str) -> SetOpReference {
+        let name_index = emitter.emit.get_atom_index(name);
+
+        //              [stack]
+
+        // FIXME: Support non-global case.
+        emitter.emit.bind_g_name(name_index);
+        //              [stack] GLOBAL
+
+        SetOpReference::new(SetOpReferenceKind::GlobalName(name_index))
+    }
+
+    #[allow(dead_code)]
+    pub fn emit_prop<F>(
+        emitter: &mut AstEmitter,
+        obj_emitter: F,
+        key: &str,
+    ) -> Result<SetOpReference, EmitError>
+    where
+        F: Fn(&mut AstEmitter) -> Result<(), EmitError>,
+    {
+        let key_index = emitter.emit.get_atom_index(key);
+
+        //              [stack]
+
+        obj_emitter(emitter)?;
+        //              [stack] OBJ
+
+        Ok(SetOpReference::new(SetOpReferenceKind::Prop(key_index)))
+    }
+
+    #[allow(dead_code)]
+    pub fn emit_elem<F1, F2>(
+        emitter: &mut AstEmitter,
+        obj_emitter: F1,
+        key_emitter: F2,
+    ) -> Result<SetOpReference, EmitError>
+    where
+        F1: Fn(&mut AstEmitter) -> Result<(), EmitError>,
+        F2: Fn(&mut AstEmitter) -> Result<(), EmitError>,
+    {
+        //              [stack]
+
+        obj_emitter(emitter)?;
+        //              [stack] OBJ
+
+        key_emitter(emitter)?;
+        //              [stack] OBJ KEY
+
+        Ok(SetOpReference::new(SetOpReferenceKind::Elem))
+    }
+}
+
+// Struct for emitting bytecode for assignment operation.
+//
+// Usage:
+//
```

Comment:
Just delete these. We don't need to duplicate code in comments.

Reply:
Do you mean all Usage comments?